matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/DefaultCryptoService.kt

```diff
@@ -636,44 +517,28 @@ internal class DefaultCryptoService @Inject constructor(
                                      callback: MatrixCallback<MXEncryptEventContentResult>) {
         // moved to crypto scope to have uptodate values
         cryptoCoroutineScope.launch(coroutineDispatchers.crypto) {
-            val userIds = getRoomUserIds(roomId)
-            var alg = roomEncryptorsStore.get(roomId)
-            if (alg == null) {
-                val algorithm = getEncryptionAlgorithm(roomId)
-                if (algorithm != null) {
-                    if (setEncryptionInRoom(roomId, algorithm, false, userIds)) {
-                        alg = roomEncryptorsStore.get(roomId)
-                    }
-                }
-            }
-            val safeAlgorithm = alg
-            if (safeAlgorithm != null) {
+            val algorithm = getEncryptionAlgorithm(roomId)
+
+            if (algorithm != null) {
+                val userIds = getRoomUserIds(roomId)
                 val t0 = System.currentTimeMillis()
                 Timber.v("## CRYPTO | encryptEventContent() starts")
                 runCatching {
-                    val content = safeAlgorithm.encryptEventContent(eventContent, eventType, userIds)
+                    preshareRoomKey(roomId, userIds)
```

Comment:
can't this be done via algorithm? and the sharing done only for megolm?

Reply:
What do you mean by that? Do you mean by `algorithm`, like it was done before where we fetch an opaque interface? If so, I would argue that the additional complexity of such an abstraction for now isn't warranted, unless we really want to support `m.olm.*` inside of rooms.

If there's a `megolm.v2` the interface won't change and the difference will be handled inside the rust code, if there's a completely different algorithm (`mls` for example) we'll likely need to change the logic around depending on the algorithm.

Reply:
The abstraction was a very legacy thing, imported from the previous SDK. I agree that this is maybe not necessary for now.